diff --git a/lib/Net/SSH2.pm b/lib/Net/SSH2.pm
index 4867f4e..3d7b23c 100644
--- a/lib/Net/SSH2.pm
+++ b/lib/Net/SSH2.pm
@@ -279,7 +279,7 @@ sub connect {
 
 sub _auth_methods {
     return {
-        ((version())[1]||0 >= 0x010203 ? (
+        (((version())[1]||0) >= 0x010203 ? (
             'agent' => {
                 ssh => 'agent',
                 method => \&auth_agent,
@@ -322,7 +322,7 @@ sub _auth_methods {
 
 sub _auth_rank {
     return [
-        ((version())[1]||0 >= 0x010203 ? ('agent') : ()),
+        (((version())[1]||0) >= 0x010203 ? ('agent') : ()),
         qw(hostbased publickey keyboard keyboard-auto password none)
     ];
 }
